### DISCUSSION TITLE
Enhancement request: allow setting & using the slug for a Matrix block

### DISCUSSION THREAD

**Discussion by adrienne** [craftcms/cms#8578] [Ideas]
Most element types, of course, have a slug or handle associated with them, but Matrix blocks, as seen in the database table `craft_elements_sites`, do not. I'd really love it if we could optionally expose a slug field for a Matrix block and then be able to address that block directly. Ideally what I want to do is something like:

```
{{ entry.myMatrix.blockSlug.fieldName }}
```

Currently I have to do something like this:
```
{% set myContentBlock = entry.matrixField.search('blockSlug::testing').first() %}
{{ myContentBlock.fieldName }}
```

Which is not a huge deal but it's kind of annoying.

**Comment by brandonkelly** [comment]
Slugs aren’t ever meant to be treated as _handles_ that should be hard-coded into the templates. At best this could be possible:

```twig
{{ entry.myMatrix.slug('my-slug').one().fieldName ?? 'default value' }}
```

**Comment by Anubarak** [comment]
You know you can use the syntax

```php
{% set myContentBlock = entry.matrixField.blockSlug('testing').first() %}
{{ myContentBlock.fieldName }}
```

As well? Because there is no way that  `{{ entry.myMatrix.blockSlug.fieldName }}` produce something useful in the end you can use a custom field with a unique validator and reproduce the behavior. Including a slug to identify them could lead to unintended usage of slugs -> might be misleading

**Comment by adrienne** [comment]
Yeah, i was really annoyed about something at the time, @brandonkelly, and you're right that I didn't think it through all the way. I do think it might be nice to expose the slugs, but it wouldn't work as a handle even so.

**Comment by narration-sd** [comment]
@adrienne  I keep seeing this pop up, and without the time to mess with it, I wonder if you (or @brandonkelly ) can get something out  of the `uri `approach, which Elements are abled to be queried on 

My briefest look and general sense of how they're employed at present suggested Entry elements and some others have `uri`s, while Matrix ones don't. But perhaps for a nice use they could, and it doesn't feel complicated, to give a pathable address.

The slug thing -- maybe you did think of it as a full path, but if only a slug-as-end, warning bells were going off about uniqueness, as I'm sure your relaxed thought would see also.

Take care, and a smile to see you around these parts
Clive
